### PR TITLE
Split up substrate-executor into polkadot-executor.

### PR DESCRIPTION
**Pull request by gavofyork**
Next part of #51 .

- Pulls apart `substrate-executor` (almost all logic) from `polkadot-executor` (just a type instantiation and tests) - (use native `impl_stubs!` macro - current a no-op, to generate the dispatch code and use it along with the wasm code).
- Refactor client into a wholly generic `substrate-client`.
- Rework `polkadot-rpc` to depend only on the generic `substrate-client` and rename to `substrate-rpc` (and rename `polkadot-rpc-servers` -> `substrate-rpc-servers`).

Still todo:

- [x] Generate dispatch code from macro `impl_stubs`.

### CODE REVIEW

File: polkadot-executor/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+extern crate polkadot_primitives as polkadot_primitives;
+extern crate ed25519;
+extern crate triehash;
+#[macro_use]
+extern crate hex_literal;
+
+use polkadot_runtime as runtime;
+use substrate_executor::{NativeExecutionDispatch, NativeExecutor};
+
+pub struct LocalNativeExecutionDispatch;
```
Comment (rphmeier) [inline polkadot-executor/src/lib.rs]: missing docs

File: polkadot-executor/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
```
Comment (rphmeier) [inline polkadot-executor/src/lib.rs]: missing docs for crate

File: polkadot-cli/src/lib.rs [hunk 3]
```diff
@@ -19,18 +19,28 @@
...
+extern crate substrate_codec as codec;
+extern crate substrate_state_machine as state_machine;
+extern crate substrate_client as client;
 extern crate substrate_primitives as primitives;
-extern crate polkadot_rpc_servers as rpc;
+extern crate substrate_runtime_io as runtime_io;
+extern crate substrate_rpc_servers as rpc;
+extern crate polkadot_primitives;
+extern crate polkadot_executor;
+extern crate polkadot_runtime;
```
Comment (rphmeier) [inline polkadot-cli/src/lib.rs]: most of these crates aren't used in this binary. why pull them in?

File: polkadot-cli/src/lib.rs [hunk 4]
```diff
@@ -53,8 +63,29 @@ pub fn run<I, T>(args: I) -> error::Result<()> where
...
+		session_length: 720,	// that's 1 hour per session.
+		sessions_per_era: 24,	// 24 hours per era.
+		bonding_duration: 90,	// 90 days per bond.
+		approval_ratio: 667,	// 66.7% approvals required for legislation.
+	};
+	let prepare_genesis = || {
+		storage = genesis_config.genesis_map();
+		let block = genesis::construct_genesis_block(&storage);
+		storage.extend(additional_storage_with_genesis(&block).into_iter());
+		use codec::Slicable;
```
Comment (rphmeier) [inline polkadot-cli/src/lib.rs]: imports at the top of the file, not scattered around!

File: polkadot-cli/src/lib.rs [hunk 5]
```diff
@@ -53,8 +63,29 @@ pub fn run<I, T>(args: I) -> error::Result<()> where
...
+		block_time: 5,			// 5 second block time.
+		session_length: 720,	// that's 1 hour per session.
+		sessions_per_era: 24,	// 24 hours per era.
+		bonding_duration: 90,	// 90 days per bond.
+		approval_ratio: 667,	// 66.7% approvals required for legislation.
+	};
+	let prepare_genesis = || {
+		storage = genesis_config.genesis_map();
+		let block = genesis::construct_genesis_block(&storage);
+		storage.extend(additional_storage_with_genesis(&block).into_iter());
```
Comment (rphmeier) [inline polkadot-cli/src/lib.rs]: redundant `into_iter`

File: polkadot-cli/src/lib.rs [hunk 6]
```diff
@@ -53,8 +63,29 @@ pub fn run<I, T>(args: I) -> error::Result<()> where
...
+		use codec::Slicable;
+		(primitives::block::Header::from_slice(&mut block.header.to_vec().as_ref()).expect("to_vec() always gives a valid serialisation; qed"), storage.into_iter().collect())
```
Comment (rphmeier) [inline polkadot-cli/src/lib.rs]: there has got to be a better way to convert in-memory than serializing and deserializing

File: polkadot-candidate-agreement/src/lib.rs [hunk 7]
```diff
@@ -314,7 +314,7 @@ impl<C: Context> SharedTable<C> {
 	/// Provide an iterator yielding pairs of (statement, received_from).
 	pub fn import_statements<I, U>(&self, iterable: I) -> U
 		where
-			I: IntoIterator<Item=(<C as TypeResolve>::SignedTableStatement, Option<C::AuthorityId>)>,
+			I: Iterator<Item=(<C as TypeResolve>::SignedTableStatement, Option<C::AuthorityId>)>,
```
Comment (rphmeier) [inline polkadot-candidate-agreement/src/lib.rs]: `IntoIterator` strictly more general and useful, why change it?

File: executor/src/native_executor.rs [hunk 8]
```diff
@@ -15,20 +15,26 @@
...
 use error::{Error, ErrorKind, Result};
-use native_runtime as runtime;
 use runtime_io;
 use state_machine::{Externalities, CodeExecutor};
 use wasm_executor::WasmExecutor;
 
 use std::panic::catch_unwind;
 
-pub struct NativeExecutor;
+pub trait NativeExecutionDispatch {
```
Comment (rphmeier) [inline executor/src/native_executor.rs]: docs missing on trait and functions

File: executor/src/native_executor.rs [hunk 9]
```diff
@@ -15,20 +15,26 @@
...
+	fn native_equivalent() -> &'static [u8];
+	fn dispatch(method: &str, data: &[u8]) -> Option<Vec<u8>>;
+}
+
+pub struct NativeExecutor<D: NativeExecutionDispatch + Sync + Send> {
+	pub _dummy: ::std::marker::PhantomData<D>,
```
Comment (rphmeier) [inline executor/src/native_executor.rs]: docs missing. also the public `_dummy` field seems strange

File: executor/src/lib.rs [hunk 10]
```diff
@@ -55,14 +51,18 @@ extern crate error_chain;
 #[cfg(test)]
 extern crate assert_matches;
 
+// TODO: move into own crate
+macro_rules! map {
```
Comment (rphmeier) [inline executor/src/lib.rs]: something that might belong in the `rstd` along with `try_opt`

File: client/src/backend.rs [hunk 11]
```diff
@@ -30,6 +30,8 @@ pub trait BlockImportOperation {
 	fn state(&self) -> error::Result<Self::State>;
 	/// Append block data to the transaction.
 	fn import_block(&mut self, header: block::Header, body: Option<block::Body>, is_new_best: bool) -> error::Result<()>;
+	/// Inject storage data into the database.
+	fn reset_storage<I: Iterator<Item=(Vec<u8>, Vec<u8>)>>(&mut self, iter: I) -> error::Result<()>;
```
Comment (rphmeier) [inline client/src/backend.rs]: maybe just `set_storage` as opposed to `reset`?

File: polkadot-cli/src/lib.rs (hunk 3 above)
Comment (gavofyork) [reply polkadot-cli/src/lib.rs]: i think most of them are used - i only managed to pull one unused one out.

File: polkadot-cli/src/lib.rs (hunk 6 above)
Comment (gavofyork) [reply polkadot-cli/src/lib.rs]: not really. right now, they're equivalent, but eventually (e.g. when we introduce `parachain_activity`) they'll only be compatible in terms of binary format. we might eventually do a specialised `From` implementation, but it's hardly critical path.

File: polkadot-candidate-agreement/src/lib.rs (hunk 7 above)
Comment (gavofyork) [reply polkadot-candidate-agreement/src/lib.rs]: accidental change.

**Comment by rphmeier** [comment]
Are there really many meaningful RPCs that can be exposed just from substrate?
Locally I am preparing a `polkadot-client-api` crate which knows which runtime functions to call to extract strongly-typed and useful data. the RPC API can also make use of that.

File: executor/src/native_executor.rs (hunk 9 above)
Comment (gavofyork) [reply executor/src/native_executor.rs]: needed to introduce the `D` type somewhere - just doing `pub struct NativeExecutor<D: NativeExecutionDispatch + Sync + Send>;` got me a compile error.

Comment (rphmeier) [reply executor/src/native_executor.rs]: Yes, what I mean is that for now the `NativeExecutor` type existing at all is unecessary

File: executor/src/lib.rs (hunk 10 above)
Comment (gavofyork) [reply executor/src/lib.rs]: yeah - out of scope for this PR but worth doing soonish.

File: client/src/backend.rs (hunk 11 above)
Comment (gavofyork) [reply client/src/backend.rs]: `set_storage` is usually used for setting a single key/value pair. i want to ensure it's known that this actually clears storage and puts the provided contents in.

**Comment by gavofyork** [comment]
- get best block (returns the header-hash)
- get header at `hash`
- get storage of `key` at `hash`
- get transaction data of `index` at `hash`

The strongly typed data (like e.g. balances) could quite reasonably be handled via secondary RPC shims.

**Comment by rphmeier** [comment]
Also cool would be the ability to evaluate a method call on the runtime at a specific header. Then we could implement a `PolkadotClient` trait for a remote endpoint as well as a reference to a client.

File: client/src/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "polkadot-cli"
```
Comment (rphmeier) [inline client/src/Cargo.toml]: accidental duplication of this `Cargo.toml` in the client source directory?

File: polkadot-executor/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+	}
+
+	fn dispatch(method: &str, data: &[u8]) -> Option<Vec<u8>> {
+		runtime::dispatch(method, data)
+	}
+}
+
+/// Creates new RustExecutor for contracts.
+pub fn executor() -> NativeExecutor<LocalNativeExecutionDispatch> {
+	NativeExecutor{ _dummy: ::std::marker::PhantomData }
```
Comment (rphmeier) [inline polkadot-executor/src/lib.rs]: missing space after `{`

File: executor/src/native_executor.rs [hunk 14]
```diff
@@ -15,20 +15,33 @@
...
+}
 
-fn safe_call<F: ::std::panic::UnwindSafe + FnOnce() -> Vec<u8>>(f: F) -> Result<Vec<u8>> {
+fn safe_call<F: ::std::panic::UnwindSafe + FnOnce() -> Option<Vec<u8>>>(f: F) -> Result<Option<Vec<u8>>> {
 	catch_unwind(f).map_err(|_| ErrorKind::Runtime.into())
 }
 
-impl CodeExecutor for NativeExecutor {
+impl<D: NativeExecutionDispatch + Sync + Send> CodeExecutor for NativeExecutor<D> {
```
Comment (rphmeier) [inline executor/src/native_executor.rs]: does it compile if we change this line to `CodeExecutor for D`?

Comment (gavofyork) [reply executor/src/native_executor.rs]: no - neither `D` nor `CodeExecutor` are local.

File: executor/src/wasm_executor.rs [hunk 15]
```diff
@@ -89,21 +91,39 @@ impl ReadPrimitive<u32> for MemoryInstance {
...
+	}
+	r
+}
+
 impl_function_executor!(this: FunctionExecutor<'e, E>,
 	ext_print_utf8(utf8_data: *const u8, utf8_len: u32) => {
 		if let Ok(utf8) = this.memory.get(utf8_data, utf8_len as usize) {
 			if let Ok(message) = String::from_utf8(utf8) {
-				info!(target: "runtime", "{}", message);
+				println!("{}", message);
```
Comment (arkpar) [inline executor/src/wasm_executor.rs]: Any good reason to revert to `println!`?

Comment (gavofyork) [reply executor/src/wasm_executor.rs]: logging doesn't work in tests, which is the primary motivation for having this.

**Pull request merged**